relay.go

```diff
@@ -0,0 +1,111 @@
+package tchannel
+
+import (
+	"errors"
+	"sync"
+)
+
+// RelayHosts allows external wrappers to inject peer selection logic for
+// relaying.
+type RelayHosts interface {
+	// Get returns the host:port of the best peer for the group.
+	Get(group string) string
+	// TODO: add MarkFailed and MarkOK to for feedback loop into peer selection.
+}
+
+type relayItem struct {
+	remapID     uint32
+	destination *Relayer
+}
+
+// A Relayer forwards frames.
+type Relayer struct {
+	sync.RWMutex
+	metrics StatsReporter
+	hosts   RelayHosts
+	items   map[uint32]relayItem
+
+	// Immutable
+	peers  *PeerList
+	conn   *Connection
+	logger Logger
+}
+
+// NewRelayer constructs a Relayer.
+func NewRelayer(ch *Channel, conn *Connection) *Relayer {
+	return &Relayer{
+		metrics: conn.statsReporter,
+		hosts:   ch.relayHosts,
+		items:   make(map[uint32]relayItem),
+		peers:   ch.Peers(),
+		conn:    conn,
+		logger:  ch.Logger(),
+	}
+}
+
+// Hosts returns the RelayHosts guiding peer selection.
+func (r *Relayer) Hosts() RelayHosts {
+	return r.hosts
+}
+
+// Relay forwards a frame.
+func (r *Relayer) Relay(f *Frame) error {
+	if f.messageType() != messageTypeCallReq {
```

Comment:
can you put a TODO in to remove a relay when we get a `messageTypeCallRes` or `callResContinue` without a continuation bit. Right now, the relayer never removes items I think. It's tricky since it's the remote side that will have to remove items.

We'll also want a TODO to remove relay items on timeout